Mock Profiles

b-quachtran opened the pull request:
- Implemented function for random user profile generation
- Updated actions.py to use generated mock profile data
- Updated amount_transferred slot sets to persist across multiple transactions
- Added date filtering to transaction search action
- Added slots to domain.yml for storing generated profile data
- Added utter_unknown_recipient to handle scenario when transfer is requested to someone not in list of known_recipients

sara-tagger commented:
Thanks for submitting a pull request 🚀 @Ghostvv will take a look at it as soon as possible ✨

akelad commented:
@b-quachtran could you add whichever email address is associated with those commits to your github profile (you can check locally with `git config --list`)? otherwise it looks like an unknown person committing. Full guide here: https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address

b-quachtran commented:
Yep, added two new intents check_recipients and check_credit_card_balance that will show user profile info. Also added conditions in the transfer/cc pay custom actions to check if the payment/transfer amount exceeds the current balance.

melindaloubser1 commented on actions/actions.py:
```diff
@@ -399,14 +422,30 @@ def validate_amount_of_money(
+            amount_of_money = float(tracker.get_slot("amount_of_money"))
+            account_balance = float(tracker.get_slot("account_balance"))
+
+            if account_balance < amount_of_money:
+                dispatcher.utter_message(template="utter_insufficient_funds")
+                return [
+                    SlotSet("PERSON", None),
+                    SlotSet("amount_of_money", None),
+                    SlotSet("confirm", None),
+                ]
```
The insufficient funds check should move to a slot validation function

b-quachtran replied:
Cool, moved the insufficient funds check to the validation function for the transfer form and added an insufficient funds check to the CC_pay validation function as well.

akelad commented:
@b-quachtran if this is ready for another review, make sure to request it from someone (prob better @erohmensing because melinda is offline for the next week)

erohmensing commented on actions/actions.py:
```diff
@@ -422,24 +462,77 @@ def name(self):
+        return []
+
+
+class ActionSessionStart(Action):
+    def name(self) -> Text:
+        return "action_session_start"
```
Actually, it may be that i'm supposed to send `/session_start` now instead. however the bot doesn't seem to like that:
```
Exception occurred while handling uri: 'http://localhost:5055/webhook'
Traceback (most recent call last):
  File "/Users/ella/miniconda3/envs/rasa/lib/python3.7/site-packages/sanic/app.py", line 976, in handle_request
    response = await response
  File "/Users/ella/RasaHQ/rasa-sdk/rasa_sdk/endpoint.py", line 102, in webhook
    result = await executor.run(action_call)
  File "/Users/ella/RasaHQ/rasa-sdk/rasa_sdk/executor.py", line 385, in run
    events = await action(dispatcher, tracker, domain)
  File "/Users/ella/RasaHQ/financial-demo/actions/actions.py", line 525, in run
    events.extend(self._slot_set_events_from_tracker(tracker))
  File "/Users/ella/RasaHQ/financial-demo/actions/actions.py", line 511, in _slot_set_events_from_tracker
    for event in tracker.applied_events()
  File "/Users/ella/RasaHQ/financial-demo/actions/actions.py", line 512, in <listcomp>
    if isinstance(event, SlotSet)
TypeError: isinstance() arg 2 must be a type or tuple of types
```

can you look into that?

b-quachtran replied:
Looks like it was a bug in my actions_session_start code, should be resolved now

erohmensing commented on actions/actions.py:
```diff
@@ -309,6 +303,24 @@ def submit(
 
+        time_range = tracker.get_slot("time")
+
+        if ("to" not in time_range) or ("from" not in time_range):
+            from_date = date.fromisoformat(time_range.split("T")[0])
+            to_date = date.today()
+        else:
+            from_date = date.fromisoformat(
+                time_range.get("from").split("T")[0]
+            )
+            to_date = date.fromisoformat(time_range.get("to").split("T")[0])
```
This confused me a little, maybe it is easier to say
```
if "to" in time_range and "from" in time_range:

else:
```
(those are the 2 different cases, right? they would be flipped in that case ofc)

b-quachtran replied:
Yep, makes sense. Conditions are flipped now

erohmensing commented on actions/actions.py:
```diff
@@ -422,24 +462,77 @@ def name(self):
+    def name(self):
+        return "action_credit_card_balance"
+
+    def run(self, dispatcher, tracker, domain):
+        cc_balance = tracker.get_slot("credit_card_balance")
+        for cc in cc_balance.keys():
+            current_balance = cc_balance[cc]["current balance"]
+            dispatcher.utter_message(
+                template="utter_credit_card_balance",
+                credit_card=cc,
+                amount_of_money=f"{current_balance:.2f}",
```
```suggestion
    def run(self, dispatcher, tracker, domain):
        credit_card_balance = tracker.get_slot("credit_card_balance")
        for card in credit_card_balance.keys():
            current_balance = credit_card_balance[card]["current balance"]
            dispatcher.utter_message(
                template="utter_credit_card_balance",
                credit_card=card,
                amount_of_money=f"{current_balance:.2f}",
```

b-quachtran commented on actions/actions.py:
```diff
@@ -378,35 +391,62 @@ def slot_mappings(self) -> Dict[Text, Union[Dict, List[Dict]]]:
+        domain: Dict[Text, Any],
+    ) -> Dict[Text, Any]:
+
+        if value and value.lower() in tracker.get_slot("known_recipients"):
```
Added title formatting to the recipient list generation in profile.py

b-quachtran commented on data/stories.md:
```diff
@@ -213,4 +213,12 @@
+* check_recipients
+    - utter_known_recipients
+
+## Show credit accounts
+* check_credit_card_balance
+    - action_credit_card_balance
```
Added this functionality in the stories and the credit card balance action, should be working now
